core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp

```diff
@@ -103,7 +103,7 @@ template <class FunctorType>
 __device__ bool cuda_inter_block_reduction(
     typename FunctorType::reference_type value,
     typename FunctorType::reference_type neutral, const FunctorType& reducer,
-    Cuda::size_type* const m_scratch_space,
+    typename FunctorType::pointer_type const m_scratch_space,
```

Comment:
> `m_scratch_space` is only used as `pointer_type` below so it seems to make most sense to just `reinterpret_cast` the pointer to that type (the size of the memory pointed to by `pointer_type` is the same for the small value case). Ultimately, we use `word_size_type` when we want to copy between different allocations but as soon as we need the actual values (for `init`, `join`, `final`) we need to cast to `pointer_type`/`value_type` anyway. Thus, it might make sense to have `scratch_data` rather as `pointer_type` in the first place and only cast it to `word_size_type` when we copy it to other allocations.

_Originally posted by @masterleinad in https://github.com/kokkos/kokkos/pull/5334#discussion_r1353257202_

Switching to `pointer_type` fits nicely here, so I'm starting with that. I could also go ahead and change it in `Kokkos_Cuda_Parallel_Team.hpp` if that makes more sense :thinking: